python/tests/catalog/test_base.py

```diff
@@ -0,0 +1,401 @@
+#  Licensed to the Apache Software Foundation (ASF) under one
+#  or more contributor license agreements.  See the NOTICE file
+#  distributed with this work for additional information
+#  regarding copyright ownership.  The ASF licenses this file
+#  to you under the Apache License, Version 2.0 (the
+#  "License"); you may not use this file except in compliance
+#  with the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing,
+#  software distributed under the License is distributed on an
+#  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+#  KIND, either express or implied.  See the License for the
+#  specific language governing permissions and limitations
+#  under the License.
+
+from typing import Dict, List, Optional, Set, Union
+
+import pytest
+
+from iceberg.catalog.base import Catalog, Identifier, Properties
+from iceberg.exceptions import (
+    AlreadyExistsError,
+    NamespaceNotEmptyError,
+    NoSuchNamespaceError,
+    NoSuchTableError,
+)
+from iceberg.schema import Schema
+from iceberg.table.base import PartitionSpec, Table
+
+
+class InMemoryCatalog(Catalog):
+    """An in-memory catalog implementation for testing purposes."""
+
+    __tables: Dict[Identifier, Table]
```

Comment:
Nit / question: I know that the usage of double underscore somewhat mangles function names to make them "private". Does that apply to class fields as well?

I've always preferred single underscore for this reason, but I'm not a huge pythonista and this is a test class. So more just for my own knowledge as well as to better understand the current "best practices".

Reply:
I noticed this too. In some ways I think the double underscore here signals "These properties don't normally exist in a non-testing Catalog implementation" but I think it'd be clearer to just directly describe that in the docstring and use property names without any underscores. You're right in that the double underscore would mangle the names, but only for a child class, something like `Foo(InMemoryCatalog)`, which isn't done anywhere here. Python would then make the property name `_InMemoryCatalog__tables` for instances of `Foo`.

I came to the same conclusion as you, this is a test class so it doesn't matter as much.

Reply:
I think a single underscore is more appropriate here. It is an internal variable of the `InMemoryCatalog`.

If we want to set this as a ClassVar, then we also have to annotation for it:
```suggestion
    _tables: ClassVar[Dict[Identifier, Table]]
```
But I'm not sure if we want to do that since we would like to instantiate a fresh InMemoryCatalog for the different tests.